adapters/repos/db/snapshot.go

```diff
@@ -0,0 +1,137 @@
+//
+// There can only be one active snapshot at a time, and creating a snapshot will
+// fail on any Index that already has an active snapshot.
+//
+// Make sure to call ReleaseSnapshot for this snapshot's ID once you have finished
+// copying the files to make sure background and maintenance processes can resume.
+func (i *Index) CreateSnapshot(ctx context.Context, id string) (*snapshots.Snapshot, error) {
+	if err := i.initSnapshot(id); err != nil {
+		return nil, err
+	}
+
+	var (
+		snap = snapshots.New(id, time.Now(), i.Config.RootPath)
+		g    errgroup.Group
+	)
+
+	for _, shard := range i.Shards {
+		s := shard
+		g.Go(func() error {
+			if err := s.createSnapshot(ctx, snap); err != nil {
+				return err
+			}
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		// TODO: restart all paused cycles here
+		//       once ReleaseSnapshot is impl,
+		//		 and store-level cycle-resume
+		//       methods are available
+		defer i.resetSnapshotState()
+		return nil, err
+	}
+
+	shardingState, err := i.marshalShardingState()
+	if err != nil {
+		// TODO: restart all paused cycles here
+		//       once ReleaseSnapshot is impl,
+		//		 and store-level cycle-resume
+		//       methods are available
+		defer i.resetSnapshotState()
+		return nil, errors.Wrap(err, "create snapshot")
+	}
+
+	schema, err := i.marshalSchema()
+	if err != nil {
+		// TODO: restart all paused cycles here
+		//       once ReleaseSnapshot is impl,
+		//		 and store-level cycle-resume
+		//       methods are available
+		defer i.resetSnapshotState()
+		return nil, errors.Wrap(err, "create snapshot")
+	}
+
+	snap.ShardingState = shardingState
+	snap.Schema = schema
+	snap.ServerVersion = config.ServerVersion
+	snap.CompletedAt = time.Now()
+
+	if err := snap.WriteToDisk(); err != nil {
+		// TODO: restart all paused cycles here
+		//       once ReleaseSnapshot is impl,
+		//		 and store-level cycle-resume
+		//       methods are available
+		defer i.resetSnapshotState()
+		return nil, err
+	}
+
+	return snap, nil
+}
+
+// ReleaseSnapshot marks the specified snapshot as inactive and restarts all
+// async background and maintenance processes. It errors if the snapshot does not exist
+// or is already inactive.
+func (i *Index) ReleaseSnapshot(ctx context.Context, id string) error {
+	i.resetSnapshotState()
+	return nil
+}
+
+func (i *Index) initSnapshot(id string) error {
+	i.snapshotStateLock.Lock()
+	defer i.snapshotStateLock.Unlock()
+
+	if i.snapshotState.InProgress {
+		return errors.Errorf(
+			"cannot create new snapshot, snapshot ‘%s’ is not yet released, this "+
+				"means its contents have not yet been fully copied to its destination, "+
+				"try again later", i.snapshotState.SnapshotID)
+	}
+
+	i.snapshotState = snapshots.State{
+		SnapshotID: id,
+		InProgress: true,
+	}
+
+	return nil
+}
+
+func (i *Index) resetSnapshotState() {
+	i.snapshotState = snapshots.State{InProgress: false}
```

Comment:
not sure if this also requires locking

Reply:
doesn't hurt to do so, will add a lock here.